File: C++/Step11-Sorting/2108.cpp
//
//  2108.cpp
//  Step11-Sorting
//
//  Created by 김예빈 on 2019. 9. 8..
//  Copyright © 2019년 김예빈. All rights reserved.
//

#include <iostream>
#include <cmath>
#include <vector>
#include <algorithm>
using namespace std;

int main(int argc, const char * argv[]) {
    cin.tie(NULL);
    ios::sync_with_stdio(false);
    
    const int MAX = 4000;
    int n, idx = 0, i;
    int mode[MAX * 2 + 1] = { 0, };
    double arithmeticalAvg = 0, medianValue, modeValue = 0, range;
    bool multi = false;
    
    cin >> n;
    
    vector<int> num(n);
    
    for(i = 0; i < n; i++) {
        cin >> num[i];
        
        arithmeticalAvg += num[i];
        mode[num[i] + MAX]++;
    }
    
    sort(num.begin(), num.end());
    
    arithmeticalAvg = arithmeticalAvg / double(n);
    arithmeticalAvg = round(arithmeticalAvg);
    
    for(i = 0; i <= MAX * 2; i++) {
        if (idx < mode[i]) {
            idx = mode[i];
            modeValue = i - 4000;
            multi = false;
        } else if(idx == mode[i] && !multi) {
            modeValue = i - 4000;
            multi = true;
        }
    }

    medianValue = num[n / 2];
    range = num[num.size() - 1] - num[0];

    cout << arithmeticalAvg << "\n" << medianValue << "\n" << modeValue << "\n" << range;
    
    return 0;
}

File: BaekJoon Online Judge/Swift/Step12-Sorting/2108.swift
//
//  2108.swift
//  Step11-Sorting
//
//  Created by Yebin Kim on 2019/12/1.
//  Copyright © 2019 김예빈. All rights reserved.
//

import Foundation

if let N = Int(readLine()!) {
    var arr = Array<Int>()
    var dic = Dictionary<Int, Int>()
    var modeValue = 0
    
    for _ in 0..<N {
        arr.append(Int(readLine()!)!)
        dic[arr.last!] = ((dic[arr.last!] != nil) ? dic[arr.last!]! : 0 ) + 1
        modeValue = (modeValue < dic[arr.last!]!) ? dic[arr.last!]! : modeValue
    }
    
    let arithmeticalAvg = round(Double(arr.reduce(0) { $0 + $1 }) / Double(arr.count))
    let medianValue = arr.sorted()[N / 2]
    let sortDic = dic.filter{ $0.value == modeValue }.sorted{ $0.key < $1.key }
    let range = arr.max()! - arr.min()!
    
    print(Int(arithmeticalAvg))
    print(medianValue)
    print(sortDic[sortDic.count > 1 ? 1 : 0].key)
    print(range)
}
